File: ios/DTO/EventPayloadDTO.swift
import CXoneChatSDK
import Foundation

struct ChatUpdatedEventDTO: Encodable {
    let state: String
    let mode: String

    init(state: ChatState, mode: ChatMode) {
        self.state = String(describing: state)
        self.mode = String(describing: mode)
    }
}

struct ThreadUpdatedEventDTO: Encodable {
    let thread: ChatThreadDTO

    init(thread: ChatThread) throws {
        self.thread = try ChatThreadDTO(thread)
    }
}

struct ThreadsUpdatedEventDTO: Encodable {
    let threads: [ChatThreadDTO]

    init(threads: [ChatThread]) {
        self.threads = threads.compactMap { try? ChatThreadDTO($0) }
    }
}

struct AgentTypingEventDTO: Encodable {
    let isTyping: Bool
    let threadId: String
    let agent: AgentDTO

    init(isTyping: Bool, agent: Agent, threadId: UUID) {
        self.isTyping = isTyping
        self.threadId = threadId.uuidString
        self.agent = AgentDTO(agent)
    }
}

struct CustomEventMessageDTO: Encodable {
    let base64: String

    init(data: Data) {
        self.base64 = data.base64EncodedString()
    }
}

struct AuthorizationChangedEventDTO: Encodable {
    let status: String
    let code: Bool
    let verifier: Bool

    init(status: String, code: Bool = false, verifier: Bool = false) {
        self.status = status
        self.code = code
        self.verifier = verifier
    }
}

struct ConnectionErrorEventDTO: Encodable {
    let phase: String
    let message: String
}

struct ErrorEventDTO: Encodable {
    let message: String
}

struct ProactivePopupActionEventDTO: Encodable {
    let actionId: String
    let action: ProactiveActionDTO

    init(actionId: UUID, payload: [String: Any]) {
        self.actionId = actionId.uuidString
        self.action = ProactiveActionDTO(actionId: actionId, payload: payload)
    }
}

struct EmptyEventDTO: Encodable {}

extension ExpoCxonemobilesdkModule {
    func sendEvent<T: Encodable>(_ name: String, dto: T) {
        let dict = (try? dto.asDictionary()) ?? [:]
#if DEBUG
        if let data = try? JSONSerialization.data(withJSONObject: dict, options: [.prettyPrinted]),
           let json = String(data: data, encoding: .utf8) {
            NSLog("[ExpoCxonemobilesdk][Event] \(name) payload:\n\(json)")
        } else {
            NSLog("[ExpoCxonemobilesdk][Event] \(name) payload serialization failed")
        }
#endif
        self.sendEvent(name, dict)
    }
}

File: android/src/main/java/expo/modules/cxonemobilesdk/dto/EventPayloadDTO.kt
package expo.modules.cxonemobilesdk.dto

import expo.modules.cxonemobilesdk.CXoneManager

data class ChatUpdatedEventDTO(val state: String, val mode: String) {
  fun toMap() = mapOf("state" to state, "mode" to mode)

  companion object {
    fun snapshot(): ChatUpdatedEventDTO {
      return ChatUpdatedEventDTO(CXoneManager.getChatStateString(), CXoneManager.getChatModeString())
    }
  }
}

data class ThreadsUpdatedEventDTO(val threads: List<ChatThreadDTO>) {
  fun toMap() = mapOf("threads" to threads.map { it.toMap() })

  companion object {
    fun from(chatThreads: List<com.nice.cxonechat.thread.ChatThread>) =
      ThreadsUpdatedEventDTO(chatThreads.map { ChatThreadDTO.from(it) })
  }
}

data class ThreadUpdatedEventDTO(val thread: ChatThreadDTO) {
  fun toMap() = mapOf("thread" to thread.toMap())

  companion object {
    fun from(thread: com.nice.cxonechat.thread.ChatThread) = ThreadUpdatedEventDTO(ChatThreadDTO.from(thread))
  }
}

data class AgentTypingEventDTO(val isTyping: Boolean, val threadId: String, val agent: AgentDTO?) {
  fun toMap() = mapOf(
    "isTyping" to isTyping,
    "threadId" to threadId,
    "agent" to agent?.toMap(),
  )
}

data class CustomEventMessageDTO(val base64: String) {
  fun toMap() = mapOf("base64" to base64)
}

data class AuthorizationChangedEventDTO(val status: String, val code: Boolean = false, val verifier: Boolean = false) {
  fun toMap() = mapOf(
    "status" to status,
    "code" to code,
    "verifier" to verifier,
  )
}

data class ConnectionErrorEventDTO(val phase: String, val message: String) {
  fun toMap() = mapOf("phase" to phase, "message" to message)
}

data class ErrorEventDTO(val message: String) {
  fun toMap() = mapOf("message" to message)
}

data class ProactiveActionDTO(
  val actionId: String?,
  val name: String?,
  val type: String?,
  val variables: Map<String, Any?>,
  val eventId: String? = null,
) {
  fun toMap() = mapOf(
    "actionId" to actionId,
    "eventId" to eventId,
    "name" to name,
    "type" to type,
    "variables" to variables,
    "content" to null, // parity with iOS payload shape (not available on Android)
  )
}

data class ProactiveActionEventDTO(val action: ProactiveActionDTO) {
  fun toMap() = mapOf(
    "actionId" to action.actionId,
    "action" to action.toMap(),
  )
}

data class EmptyEventDTO(val placeholder: Boolean = true) {
  fun toMap() = emptyMap<String, Any?>()
}
